### PR TITLE
add no-identical-functions rule

### LINKED ISSUE #17: Functions should not have identical implementations
https://jira.sonarsource.com/browse/RSPEC-4144

### PR DESCRIPTION
**Pull request by stas-vilchik**
fixes #17

### CODE REVIEW

File: src/rules/no-identical-functions.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        visitFunction(node as estree.ArrowFunctionExpression);
+      },
+
+      "Program:exit"() {
+        processFunctions();
+      },
+    };
+
+    function visitFunction(node: estree.Function) {
+      if (isBlockStatement(node.body) && isBigEnough(node.body)) {
```
Comment (vilchik-elena) [inline src/rules/no-identical-functions.ts]: @stas-vilchik why did you keep `isBlockStatement`?

File: src/utils/locations.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    } else {
+      const token = context.getSourceCode().getFirstToken(fn);
+      location = token && token.loc;
+    }
+  } else if (fn.type === "FunctionExpression") {
+    const parent = getParent(context);
+    if (parent && parent.type === "MethodDefinition") {
+      location = parent.key.loc;
+    } else {
+      const token = context.getSourceCode().getFirstToken(fn);
```
Comment (vilchik-elena) [inline src/utils/locations.ts]: @stas-vilchik can be `async`

File: src/utils/locations.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  if (fn.type === "FunctionDeclaration") {
+    // `fn.id` can be null when it is `export default function` (despite of the @types/estree definition)
+    if (fn.id) {
+      location = fn.id.loc;
+    } else {
+      const token = context.getSourceCode().getFirstToken(fn);
+      location = token && token.loc;
+    }
+  } else if (fn.type === "FunctionExpression") {
+    const parent = getParent(context);
```
Comment (vilchik-elena) [inline src/utils/locations.ts]: @stas-vilchik It doesn't work for method,`context` here is at `Program:exit`

File: tests/rules/no-identical-functions.test.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    {
+      // different kinds of functions
+      code: `
+      function foo() {
+        console.log("Hello");
+        console.log("World");
+        return 42;
+      }
+
+      let funcExpression = function () {
```
Comment (vilchik-elena) [inline tests/rules/no-identical-functions.test.ts]: @stas-vilchik `// Noncompliant` here and for arrow function

File: src/rules/no-identical-functions.ts (hunk 1 above)
Comment (stas-vilchik) [reply src/rules/no-identical-functions.ts]: to check that a function has at least one statement

File: tests/rules/no-identical-functions.test.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        return 42;
+      }
+
+      function bar() {
+        console.log("Hello");
+        return 42;
+      }`,
+    },
+    {
+      // ignore not block statement body
```
Comment (vilchik-elena) [inline tests/rules/no-identical-functions.test.ts]: @stas-vilchik this case is mistyped? there are different braces

Comment (stas-vilchik) [reply tests/rules/no-identical-functions.test.ts]: that's the point

**Pull request merged**